### ISSUE TITLE
Provide ``pre-commit`` hooks even without having to use a system hook

### ISSUE THREAD

**Issue by Pierre-Sassoulas** [PyCQA/pylint#5293]
### Current problem

Pylint need the code dependencies to be installed to be able to analyze the code. So it need to be aware of the dependencies it needs if we want to be able to install it in an isolated virtualenv environment for pre-commit. 

Right now this is not the case so the pre-commit hook file that is in the repository is miseleading.

### Desired solution

Find a way to make the dependencies known to pylint (new option ?) so it's installed by pre-commit in the isolated venv. Or make pylint detect the dependencies from the setup.py / setup.cfg / pyproject.toml and install them ? I'm not really comfortable with installing dependencies at run time, and I can't think of a better design myself, so I'm open to suggestions.

### Additional context

https://github.com/PyCQA/pylint/pull/5292, and #4505

**Comment by adamchainz** [comment]
pre-commit does not like dependencies changing at runtime, as it tries to build immutable per-tool environments based on its own configuration only.

pre-commit has the `additional_dependencies` option to add extra packages into the environment, which can use `-r requirements.txt` etc., but that breaks its environment caching. It won't reinstall when `requirements.txt` changes, which can cause developers’ setups to become out of sync.

If pylint could be pointed to a second virtualenv where dependencies are installed, that could work. (I presume pylint only uses AST analysis of dependencies, rather than importing them.)

**Comment by ssbarnea** [comment]
@adamchainz describes a relatively well know issue that affects linters as `pylint` and `mypy` and I am aware of these possible workarounds:

* Duplicate each dependency needed inside `additional_dependencies` and keep using it. They can get out of sync but is not a big deal. The linter will fail if you add new dependency to the project so you will know to adjust its list. When removing dependencies from project you may forget to remove them from linter but that didn't happen to me yet, mainly because when I drop a dependency I search the codebase for module name, and that will uncover the extra location where you have to remove it from.
* Use local system hooks, usually configured to run as `manual` stage so people can still use pre-commit directly and avoiding breaking on pre-commit.ci service. I dislike this setup as I value a lot both ways that are crippled by using it.
* Disable pylint checks that need imports like `import-error`, discouraged too for obvious reasons.

I wrongly used the `-r requirements.txt` trick or even `-e .` in the past to force pre-commit to install dependencies from outside its config, just to discover that this caused cached venv to get out of sync and forcing me or others to have to reset entry pre-commit cache to make it work again. I could say that I learnt the hard way to avoid this trick.

**Comment by Pierre-Sassoulas** [comment]
(Right now, I'm creating a venv then using a dev extra require with pylint (pytest, pytest-cov...) and doing a ``pip install -e ".[dev"]"`` personally). 

I think a little duplication in ``additional_dependencies``makes sense and beat having to think of installing in venv all the time or adding extra require to the packaging (and it works without packaging too). @ssbarnea I would appreciate if you add your insight to the documentation and to the example pre-commit configuration in a MR.

**Comment by DanielNoord** [comment]
@ssbarnea I think your comment above make total sense. Are you still up for creating a PR with some additional comments in those files?

**Comment by ssbarnea** [comment]
Yeah, I could document the first bullet. I am not inclined to mention the other two options as they are to be discouraged due to the unpleasant side-effects or limited functionality.

**Comment by jacobtylerwalls** [comment]
@Pierre-Sassoulas I have a confession to make. I haven't been using pre-commit when contributing to pylint for a long time, at least a year.

pre-commit runs fine like `precommit run -a`, but never with git commit, always failing with:
```shell
pylint...................................................................Failed
- hook id: pylint
- exit code: 1

Executable `pylint` not found
```

Same whether dev'ing in a venv, or outside of one, like so:
```sh
% which pylint
/Library/Frameworks/Python.framework/Versions/3.9/bin/pylint
```

Since `language: system` means `you're on your own`, I won't open an issue against pre-commit until I understand more. I switched to `language: python` and reproduced exactly the same issue, though. Any advice for me (or have you seen this before)?

**Comment by Pierre-Sassoulas** [comment]
Personally I'm using an editable install of pylint in a venv and the venv is always activated. (The message ``Executable `pylint` not found`` would indicate not being in the venv in my setup). So to be clear what I'm doing:

```bash
python3 -m venv venv
source venv/bin/activate
pip3 install -e ".[spelling]"
pre-commit install
```

But more and more I'm thinking that pylint is not a tool for pre-commit hook and more for continuous integration check:
- It's slow
- It's okayish even when the pre-commit hook is executed on the modified files only because of astroid import time being outrageous (> 0.5s)
- It requires a fully setup env (possibly with properly compiled everything) and editable installs are out of the question in pre-commit it seems, so even for pure python projects it means duplicating everything in ``additional_dependencies``.

So being able to do ``git rebase -i upstream/main --exec=pre-commit run --all-files`` is really annoying and daily use is slightly annoying. I'm wondering if we should really advertise pylint as a ``pre-commit`` hook at this point.

**Comment by jacobtylerwalls** [comment]
With apologies for continuing to slightly hijack this thread, just mentioning that I repeated the same commands you provided and reproduced the exact same issue. My $PATH must be more convoluted (?)

**Comment by mbyrnepr2** [comment]
I'm not sure anymore why [this documentation ](https://pylint.readthedocs.io/en/stable/user_guide/installation/pre-commit-integration.html) mentions: 
_the hook only works with a local installation of pylint (in your environment)_
and also uses the `- repo: local` configuration.

The first bullet point by @ssbarnea above is what I've tried and it works well without using the local hook approach.
```python
-   repo: https://github.com/pylint-dev/pylint
    rev: v3.0.0a6
    hooks:
    - id: pylint
      args: ["--rcfile=pylintrc"]
      files: src
      additional_dependencies: ["requests", "PyYAML"]
```

Similar approach [here](https://github.com/psf/black/blob/main/.pre-commit-config.yaml#L46) by black.
I could be missing something, but do we simply need to update that doc to show how the hook can be used the non-local way, since that seems the easiest way for users to interact with it?
